Handle crossorigin in Link

nmvk opened the pull request:
Implemented Step three and handled step four of obtain the resource part
of 4.2.4 The link element.
Link to spec : https://html.spec.whatwg.org/multipage/semantics.html#concept-link-obtain

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

<!-- Either: -->
- [X] There are tests for these changes OR

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14940)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @fitzgen: components/script/dom/htmllinkelement.rs, components/script/dom/webidls/HTMLLinkElement.webidl, components/script/stylesheet_loader.rs
 * @kichjang: components/script/dom/htmllinkelement.rs, components/script/dom/webidls/HTMLLinkElement.webidl, components/script/stylesheet_loader.rs

jdm commented on components/script/dom/htmllinkelement.rs:
```diff
@@ -377,6 +387,32 @@ impl HTMLLinkElementMethods for HTMLLinkElement {
     // https://html.spec.whatwg.org/multipage/#dom-link-target
     make_setter!(SetTarget, "target");
 
+    // https://html.spec.whatwg.org/multipage/#dom-link-crossorigin
```
This should be using an enumerated getter instead: https://dxr.mozilla.org/servo/rev/124301cf484818a87ab0b3e0c4ef34645be66174/components/script/dom/htmlimageelement.rs#296

mrnayak replied:
Using `enumerated_getter` always returns default crossorigin value as `anonymous`. This value is returned even if request is not a valid crossorigin request. This breaks many test case.

jdm replied:
You're right. Could we share this code with htmlscriptelement.rs instead of duplicating it, in that case?

mrnayak replied:
I have fixed this.

jdm commented on components/script/dom/htmllinkelement.rs:
```diff
@@ -377,6 +387,32 @@ impl HTMLLinkElementMethods for HTMLLinkElement {
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-link-crossorigin
+    fn SetCrossOrigin(&self, value: Option<DOMString>) {
```
This can use `make_setter`: https://dxr.mozilla.org/servo/rev/124301cf484818a87ab0b3e0c4ef34645be66174/components/script/dom/htmlimageelement.rs#298

jdm replied:
My mistake; it can't. Let's add another `set_cross_origin_attribute` helper we can call from both HTMLLinkElement and HTMLScriptElement.

jdm commented on components/script/dom/htmllinkelement.rs:
```diff
@@ -235,6 +237,14 @@ impl HTMLLinkElement {
         };
 
+        // Step 3
+        let cors_setting = match self.GetCrossOrigin() {
```
Let's make a helper function for this and share it with htmlscriptelement.rs.

mrnayak replied:
I have added a method in `request.rs` which given a `&str` returns `CorsSettings`. I am not sure if this is good refactoring. Other option could be return `CorsSettings` given `DomString` but I dont know what is the right location for such utility.

jdm replied:
I propose the following in element.rs:
```rust
fn reflect_cross_origin_attribute(element: &Element) -> Option<DOMString> {
    // body of HTMLScriptElement::GetCrossOrigin
}

fn cors_setting_for_element(element: &Element) -> Option<CorsSetting> {
    reflect_cross_origin_attribute(element).map(|attr| {
        // body of CorsSetting::from_str_ref
    })
}
```
Then these can be called by both HTMLLinkElement and HTMLScriptElement as necessary.

jdm commented on components/script/stylesheet_loader.rs:
```diff
@@ -244,6 +255,6 @@ impl<'a> StylesheetLoader<'a> {
 impl<'a> StyleStylesheetLoader for StylesheetLoader<'a> {
     fn request_stylesheet(&self, import: &Arc<RwLock<ImportRule>>) {
-        self.load(StylesheetContextSource::Import(import.clone()), "".to_owned())
+        self.load(StylesheetContextSource::Import(import.clone()), None, "".to_owned())
```
Let's add a TODO comment about whether we should use the original loader's CORS setting. I don't find any direction from the specification about it.

jdm reviewed (changes requested):
I'm glad to see more tests passing. I've made a few suggestions for ways to avoid duplicating code.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14911) made this pull request unmergeable. Please resolve the merge conflicts.

nmvk committed:
Handle crossorigin in link and refactor crossorigin handling

Implemented Step three and handled step four of obtain the resource part
of 4.2.4 The link element.
Link to spec : https://html.spec.whatwg.org/multipage/semantics.html#concept-link-obtain

Refactored crossOrigin handling in HTMLScriptElement, HTMLImageElement

mrnayak commented:
@jdm Thanks, I have addressed review comments. Additionally I refactored `HTMLImageElement` to use this updated change.

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 3d9e44a has been approved by `jdm`

<!-- @bors-servo r=jdm 3d9e44a8c497acd52b428dd02062a9c55a2efafc -->

bors-servo commented:
:hourglass: Testing commit 3d9e44a with merge 9afd960...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/4888), [arm32](http://build.servo.org/builders/arm32/builds/5051), [arm64](http://build.servo.org/builders/arm64/builds/5096), [linux-dev](http://build.servo.org/builders/linux-dev/builds/5196), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1868), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1878), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/5164), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/5206), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1870), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1913), [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/332), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/334)
